docs/architecture/adr-110-e2e-abci++.md

```diff
@@ -0,0 +1,267 @@
+- `test/e2e/pkg/grammar/lexer`,
+- `test/e2e/pkg/grammar/parser`,
+- `test/e2e/pkg/grammar/sppf`,
+- `test/e2e/pkg/grammar/token`.
+
+Apart from this auto-generated code, we implemented `GrammarChecker` abstraction
+which knows how to use the generated parser and lexer to verify whether a
+specific execution (list of ABCI++ calls logged by the Application while the
+testnet was running) respects the ABCI++ grammar. The implementation and tests 
+for it are inside `test/e2e/pkg/grammar/checker.go` and 
+`test/e2e/pkg/grammar/checker_test.go`, respectively. 
+
+How the `GrammarChecker` works is demonstrated with the test `TestABCIGrammar`
+implemented in `test/e2e/tests/abci_test.go` file. 
+
+```go
+func TestABCIGrammar(t *testing.T) {
+	m := fetchABCIRequestsByNodeName(t)
+	checker := grammar.NewGrammarChecker(grammar.DefaultConfig())
+	testNode(t, func(t *testing.T, node e2e.Node) {
+		reqs := m[node.Name]
+		_, err := checker.Verify(reqs)
+		if err != nil {
+			t.Error(err)
+		}
+	})
+}
+```
+
+Specifically, the test first fetches all ABCI++ requests and creates a `GrammarChecker` object. Then for each
+node in the testnet it checks if a specific set of requests respects the ABCI++ 
+grammar by calling `checker.Verify(reqs)` method. If this method returns an error, the specific execution does not respect the grammar. 
+
+The `Verify()` method is shown below. 
+```go
+
+func (g *GrammarChecker) Verify(reqs []*abci.Request) (bool, error) {
+	var r []*abci.Request
+	var n int
+	if g.cfg.FilterLastHeight {
+		r, n = g.filterLastHeight(reqs)
+		if n != 0 {
+			debugMsg := fmt.Sprintf("Last height filtered, removed last %v abci calls out of %v.\n", n, len(reqs))
+			g.logger.Debug(debugMsg)
+		}
+	}
+	s, _ := g.GetExecutionString(r)
+	return g.VerifyExecution(s)
+}
+```
+
+It takes a list of requests and does the following things.
+- Filter the last height. Basically, it removes all ABCI++ requests after the 
+last `Commit`. Function `fetchABCIRequestsByNodeName()` can be called in the middle of the height. As a result, the last height may be incomplete, and 
+the parser may return an error. The simple example here is that the last 
+request fetched via `fetchABCIRequestsByNodeName()` is `PrepareProposal`; however, `ProcessProposal` happens after 
+`fetchABCIRequestsByNodeName()` was invoked. Consequently, the parser
+will return an error that `ProcessProposal` is missing, even though the `ProcessProposal` 
+will happen after.  
+- Generates an execution string by replacing `abci.Request` with the 
+corresponding terminal from the grammar. This logic is implemented in
+`GetExecutionString()` function. This function receives a list of `abci.Request` and generates a string where every request the grammar covers 
+will be replaced with a corresponding terminal. For example, `abci.Request_PrepareProposal` is replaced with `<PrepareProposal>`. If the request is not covered 
+by the grammar, it will be ignored. 
```

Comment:
> If the request is not covered by the grammar, it will be ignored.

I guess this is because of `ExtendVote` and `VerifyVoteExtension`. Are you planning to reject any terminal not defined in the grammar once you upgrade the grammar (and the ADR, and the code) to support vote extensions?

Reply:
Yes, for example, we will ignore the `Info` abci request.